live-build/misc/live-build-hooks/90-linux-migration-artifact.binary

```diff
@@ -0,0 +1,164 @@
+# for those. /opt/delphix is always created as part of the
+# appliance. Thus, checking if it exists is a quick way for us to
+# distinguish between variants that have the appliance from
+# the ones that don't.
+#
+if [[ ! -d binary/opt/delphix ]]; then
+	echo "This artifact does not contain the appliance." \
+		" Skipping migration image..."
+	exit 0
+fi
+
+mkdir $DEPOT_DIRECTORY
+
+#
+# Generate archive for os root directory.
+#
+# Note: We temporarily change directory in the subshell (we move to the
+# "binary" directory) because find prints everything as a relative
+# path to the current directory and we don't want the generated archive
+# to have "binary" as its top-level directory.
+#
+(
+	cd binary
+	find ./* -print | cpio -oc
+) >$DEPOT_DIRECTORY/os-root.cpio
+
+#
+# Copy all migration scripts including dx_prepare which is needed by
+# dx_unpack.
+#
+cp migration-scripts/* $DEPOT_DIRECTORY
+
+#
+# There may be a version.info file in the current directory already.
+# That version.info is Linux-specific and does not necessarily have
+# all the fields that the illumos upgrade process requires. Thus, we
+# create our own special version info directly in the depot directory
+# with all the fields expected by dx_unpack.
+#
+{
+	#
+	# Note that both the following fields are bogus and supplied
+	# solely for making dx_unpack happy.
+	#
+	echo "DLPX_DATE=bogus"
+	echo "DLPX_OS_VERSION=Linux"
+
+	#
+	# DLPX_MIN_VERSION is specified for ensuring that we are
+	# migrating from the right version.
+	#
+	echo "DLPX_MIN_VERSION=5.3.0.0"
+
+	#
+	# DLPX_VERSION is set explicitly to match the version of the
+	# virtualization appliance, which is what upgrades in Linux
+	# do and therefore we try to stay consistent as this field
+	# really can be anything (e.g. a date, a point-version, a
+	# nickname, etc.)
+	#
+	# shellcheck disable=SC2016
+	echo "DLPX_VERSION=$(chroot binary dpkg-query \
+		-Wf '${Version}' delphix-virtualization)"
+} >$DEPOT_DIRECTORY/version.info
+
+#
+# Generate hashes file.
+#
+(
+	cd binary
+	find ./* -type f -print0 | xargs -0 sha256sum
+) >$DEPOT_DIRECTORY/os-root.hashes
+(
+	cd $DEPOT_DIRECTORY
+	sha256sum ./* >hashes
+)
+
+#
+# DLPX_SIGNING_LOGIN is used for signing our migration image.
+# The signature is later unpacked and verified on the VM
+# performing the migration. If DLPX_SIGNING_LOGIN has not been
+# specified by the user when running this hook we skip generating
+# the signature for this image.
+#
+if [[ -n "${DLPX_SIGNING_LOGIN:-}" ]]; then
+	#
+	# Generate depot/hashes.sig[.<release>]
+	#
+	# Assumption: we always migrate from version 5.3
+	#
+	UPGRADE_VERSION="5.3"
+	DLPX_SIGNING_URL="$DLPX_KEY_URL/upgrade/keyVersion/$UPGRADE_VERSION/sign"
+
+	set -o pipefail
+	curl -s -f -H "Content-Type: application/json" \
+		-u "$DLPX_SIGNING_LOGIN" "$DLPX_SIGNING_URL" -d @- <<-EOF |
+			{"data": "$(base64 -w 0 $DEPOT_DIRECTORY/hashes)"}
+		EOF
+		jq -r .signature |
```

Comment:
did you test this part of the code? I'm surprised we don't have to add `jq` to the Docker image...

Reply:
We don't have to add it because it is already there :-)
```
...(from Dockerfile) ....
RUN \
  apt-get update && \
  apt-get install -y software-properties-common && \
  apt-add-repository -y ppa:ansible/ansible && \
  apt-get update && \
  apt-get install -y \
      ansible \
      aptly \
      awscli \
      bc \
      coreutils \
      equivs \
      gdisk \
      git \
      java-package \
      jq \   <-------
....
```

```
$ git grep jq
docker/Dockerfile:      jq \
live-build/misc/live-build-hooks/90-linux-migration-artifact.binary:            jq -r .signature |
live-build/misc/live-build-hooks/92-ova-machine-image.binary:   jq '.["actual-size"]')
live-build/misc/live-build-hooks/92-ova-machine-image.binary:   jq '.["virtual-size"]')
```

Reply:
Sorry. Clearly I’m blind; I looked the Dockerfile before leaving the comment, and somehow missed that. Thanks.